Committing UXUI-212 for Credit Card Expiration

manifestinteractive opened the pull request:
## Credit Card Expires Component

New component for https://holidayextras.jira.com/browse/UXUI-212  
Passes `npm test` and `npm run lint`

#### Attributes

`defaultValue` String - ( default empty ) Initial Value of Credit Card Expiration Date

## Example

```js
var cc = {
  expires: '04/15'
};

var example = (
  <UIToolkit.CreditCardExpire defaultValue={cc.expires} />
);

React.render(example, mountNode);
```

#### Screenshots / Screencast
![screenshot](http://i.imgur.com/Hp4fQXQ.png)

---
#### Developer Definition of Done/Quality Checklist (for PR author to complete BEFORE code review):
- [x] I have checked the [contributing document](../blob/master/CONTRIBUTING.md) and I'm happy for this to be reviewed.

#### Software Engineer or Developer review:
- [ ] I’ve witnessed the work behaving as expected (this could be on the authors machine or screencast).
- [ ] I’ve checked for coding anti-patterns.
- [ ] I’ve checked for appropriate test coverage.
- [ ] I’ve run all the tests locally and they pass.

#### Software Engineer or project guru review:
- [ ] I’ve witnessed the work behaving as expected (this could be on the authors machine or screencast).
- [ ] I’ve checked for coding anti-patterns.
- [ ] I’ve checked for appropriate test coverage.
- [ ] I’ve run all the tests locally and they pass.

cameronviner commented:
Taking DEV

cameronviner commented on src/components/credit-card-expire/code/views/creditCardExpireView.jsx:
```diff
@@ -0,0 +1,74 @@
+
+  updateValue: function(expireDate) {
+
+    if(typeof expireDate === 'undefined')
```
Any chance we could change it so that the opening curly brace is on the same line as the if?

cameronviner commented:
Hey @manifestinteractive 

This doesn't seem to be working for me, if I enter 10/15 for instance it doesn't seem to work: https://cloudup.com/cDqi2yyCtXR

This seems to be happening with any month 10 and above.

manifestinteractive commented:
oops, likes like thats a bug with the leading 10, I will fix this.  It 
was intended for people entering in dates like 915 to become 09/15 
likewise 9/15 being entered should still convert to 09/15.

I will get this out ASAP ( and fix the bracket placement ).

We may want to update the "npm run lint" for this kind of stuff though 
if we are worried about brace placement, that can be checked with a 
linter, which is what I was using for syntax validation.

> Cameron Viner <mailto:notifications@github.com>
> May 28, 2015 at 10:09 AM
>
> Hey @manifestinteractive <https://github.com/manifestinteractive>
>
> This doesn't seem to be working for me, if I enter 10/15 for instance 
> it doesn't seem to work: https://cloudup.com/cDqi2yyCtXR
>
> This seems to be happening with any month 10 and above.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/holidayextras/ui-toolkit/pull/25#issuecomment-106335099>.
>

jodiedoubleday commented on src/components/credit-card-expire/code/views/creditCardExpireView.jsx:
```diff
@@ -0,0 +1,74 @@
+      var month = parseInt(dateParts[0], 10) || '';
+      var year = parseInt(dateParts[1], 10) || '';
+
+      if(month > 12)
```
star `{` needs to be on same line as `if`

jodiedoubleday commented on src/components/credit-card-expire/code/views/creditCardExpireView.jsx:
```diff
@@ -0,0 +1,74 @@
+      var year = parseInt(dateParts[1], 10) || '';
+
+      if(month > 12)
+      {
```
`{` needs to be on same line as `if`

jodiedoubleday commented on src/components/credit-card-expire/code/views/creditCardExpireView.jsx:
```diff
@@ -0,0 +1,74 @@
+      {
+        month = 12;
+      }
+      else if(month > 0 && month < 12)
```
as above

jodiedoubleday commented on src/components/credit-card-expire/code/views/creditCardExpireView.jsx:
```diff
@@ -0,0 +1,74 @@
+        month = '0' + month;
+      }
+
+      if(month && year)
```
as above

manifestinteractive commented:
__Addresses the following:__

1.  Merged in master and addressed merge conflicts
2.  Updated Unit Tests
3.  Fixed issue with month / year formatting autocompletion
4.  Updated Syntax Format as requested
5.  npm test && npm run lint pass

jodiedoubleday commented:
This still has an issue where i type `0115` and it fills out `01/15` (as expected) but then i realise i've made a mistake and try to change the `01` to `11` and it goes a bit crazy.
Screencast here - http://quick.as/9ze7hp6x4

manifestinteractive commented:
Swapped out custom Validation for third-party Payment library.  Added error class when invalid input as well.

hx-markterry commented:
This component is going to need i18n support, as it contains a few hardcoded english phrases

manifestinteractive commented:
@hx-markterry I will make these properties you can pass in with the English text as default, rather than hard code.

manifestinteractive commented:
@hx-markterry Updated to make errorMessage a property that can be passed in and defaults to english included in component.  Can you clarify other places where you see hard coded English?  Everything else should have been only default properties ( meaning they can be overloaded just by setting the components property when creating it ).

hx-markterry commented:
Hard coded strings all look good now, thanks.

hx-markterry commented:
I checked this out on my machine, using the docs I can set the expiry date to just `12/` and it does not give an error.

lukehansell-hx commented:
To make this component more re-usable could we remove the "Expires"? It occurs to me that this is just a date input formatted to 'MM/YY' with a label and could as easily be used for credit card start dates. We lose nothing here by increasing the component scope to this level...

manifestinteractive commented:
@lukefrake The "Expires" is a label that is added only if you set it.  If you look at the example, the label property is set to "Expires" and that is the only reason it shows up.  You can literally make this label whatever you want.  Or don't use it at all.

As seen in the ./ui-toolkit/src/components/credit-card-expire/code/views/creditCardExpireView.jsx file, there is a condition that only adds that label if you set it, and uses whatever value as the text for that label.

Make sense ?

lukehansell-hx commented:
I think that's my point, if this can be a date for expiry or start (based only on what is passed as the label text) then it shouldn't be called "Credit Card Expire" (with CreditCardExpire.jsx etc) but instead "Credit Card Date". 

This component is agnostic of what the date being entered is for and should be named as such otherwise the use context is distorted.

manifestinteractive commented:
@lukefrake I thought you were using the quotes to symbolize a text string, so was assuming you were talking about the label.  These variable/file names never get touched by a normal user of this component. Please let me know if this ticket can't get merged because of variable/file names and I will make the requested update.

lukehansell-hx commented:
It's exported as CreditCardExpire on the UIToolkit object as well as in the documentation write-up. To keep consistency within the toolkit, be declarative of the purpose and ease of readability etc for developers maintaining the toolkit the files and variables should probably be renamed.

manifestinteractive commented:
@lukefrake OK, I will make this change once I get the green light that there is nothing else that needs changing, otherwise I should wait until all feedback is complete before making this change.  Since you were the only one that caught this, can you tell me what you prefer the naming convention be?  I see "Credit Card Date" but just want to confirm to prevent any further delay getting this out.

lukehansell-hx commented:
Personally I believe CreditCardDate gives a clear indication of the use of the element. If anyone else would like to weigh in I'm open to other suggestions?

manifestinteractive commented:
@lukehansell-hx Updated all references to "expires" in this plugin to "date" throughout the entire component.  Updated demo in docs, verified tests passed & merged in latest master for ease of pressing that "Merge pull request" button that's just right down there... ;)

lukehansell-hx commented on docs/src/Components.jsx:
```diff
@@ -92,6 +92,21 @@ var Components = React.createClass({
+          <li><code>placeHolder</code> [optional] String - Default  'MM / YY',</li>
+          <li><code>name</code> [optional] String - Default 'credit-card-date',</li>
+          <li><code>id</code> [optional] String - Default 'credit-card-date',</li>
+          <li><code>errorMessage</code> [optional] String - Default 'Invalid Date'</li>
```
`label` is an attribute not listed here

lukehansell-hx commented on src/components/credit-card-date/code/views/creditCardDateView.jsx:
```diff
@@ -0,0 +1,67 @@
+
+  getDefaultProps: function() {
+    return {
+      formatInput: false,
```
is it me or is this not actually used anywhere?

lukehansell-hx commented:
entering a date such as `03/19` or `03/ 19` (basically anything without the whitespace surrounding the slash) is invalid. Should we loosen the validation here to accept (and possibly reformat) these cases. We could possibly also do with accepting 0319 as MMYY (but that isn't necessary, I wonder if we could get some stats on what people try to enter into our expiry date boxes currently?).

lukehansell-hx commented on src/components/credit-card-date/__tests__/creditcard-test.js:
```diff
@@ -0,0 +1,25 @@
+    assert.isDefined(renderedCreditCardLabel);
+  });
+
+});
```
can we test:
- not having a label
- having an error
- not having an error
- defining custom values for attributes
- checking custom attributes
- event handlers
...

I'm not sure if it's possible to test all of the above and more, but there seems to be a bit of functionality in this component that isn't currently tested. I feel that since these components are starting to go into production we need to vastly increase our test coverage and a new component seems like the perfect place to start. Using [`React.Addons.TestUtils`](https://facebook.github.io/react/docs/test-utils.html#simulate) we should be able to simulate each of these circumstances.

lukehansell-hx commented on src/components/credit-card-date/code/views/creditCardDateView.jsx:
```diff
@@ -0,0 +1,67 @@
+
+    if(expires.length > 0){
+      isValid = false;
+      var date = expires.split(' / ');
```
this doesn't account for "0915" or "09 /15" etc... could this be loosened as I think it's safe to assume that not all users will enter card details with the correct whitespacing.

lukehansell-hx commented on src/components/credit-card-date/scripts/build-dev.sh:
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
```
this file can be removed as "dev" builds are no longer supported, use docs instead

lukehansell-hx commented on src/components/credit-card-date/scripts/build-dist.sh:
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
```
do we require this file?

lukehansell-hx commented on src/components/credit-card-date/scripts/build-dist.sh:
```diff
@@ -0,0 +1,7 @@
+
+echo "Building package for demo. Variant: '$VARIANT', language: '$TEMPLATE_LANG'."
+
+browserify -t [ redirectify --dir "$VARIANT" ] -t reactify index.js --standalone ui-component-creditcard > dist/ui-component-creditcard-standalone.js
```
if we're keeping this then we should probably change `ui-component-creditcard` to `ui-component-credit-card-date`

lukehansell-hx commented on src/components/credit-card-date/code/views/creditCardDateView.jsx:
```diff
@@ -0,0 +1,78 @@
+
+  handleChange: function(e){
+
+    var elm = document.querySelector('.credit-card-date-input');
```
This doesn't work if we have multiple card date inputs on the page such as "start date" and "expiry date". The second input would never be formatted. Does e not provide us the active element?

manifestinteractive replied:
changing to:

```javascript
if( !this.formatting && e.target){
    this.formatting = true;
    Payment.formatCardExpiry(e.target);
}
```
